### PR TITLE
Consolidate Cubeviz parser conversions

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is leverages the following specutils PRs

* https://github.com/astropy/specutils/pull/1181
* https://github.com/astropy/specutils/pull/1184

to simplify Cubeviz parser unit conversion logic. I also refactored it a bit to streamline the flow because things are looking a bit chaotic with different people inserting parts of this and that in the parser.

As a result, I might have also fixed bugs in some uncommon cases that we test for but users do not usually encounter.

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-4788)

### CODE REVIEW

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 1]
```diff
@@ -522,15 +521,15 @@ def _parse_ndarray(app, file_obj, data_label=None, data_type=None,
...
-        flux = flux << u.count
+        flux = flux << (u.count / PIX2)
 
     meta = standardize_metadata({'_orig_spatial_wcs': None})
     s3d = Spectrum1D(flux=flux, meta=meta)
 
     # convert data loaded in flux units to a per-square-pixel surface
     # brightness unit (e.g Jy to Jy/pix**2)
-    if not check_if_unit_is_per_solid_angle(s3d.unit):
-        file_obj = convert_spectrum1d_from_flux_to_flux_per_pixel(s3d)
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:533]: I think assigning `file_obj` here is actually a bug as `file_obj` does not get used after this line.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 2]
```diff
@@ -556,12 +555,7 @@ def _parse_gif(app, file_obj, data_label=None, flux_viewer_reference_name=None):
     flux = np.rot90(np.moveaxis(flux, 0, 2), k=-1, axes=(0, 1))
 
     meta = {'filename': file_name, '_orig_spatial_wcs': None}
-    s3d = Spectrum1D(flux=flux * u.count, meta=standardize_metadata(meta))
-
-    # convert data loaded in flux units to a per-square-pixel surface
-    # brightness unit (e.g Jy to Jy/pix**2)
-    if not check_if_unit_is_per_solid_angle(s3d):
-        file_obj = convert_spectrum1d_from_flux_to_flux_per_pixel(s3d)
+    s3d = Spectrum1D(flux=flux * (u.count / PIX2), meta=standardize_metadata(meta))
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:558]: Really, baby shark does not need so many checks, just assign it whatever unit you want here.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 3]
```diff
@@ -580,67 +574,3 @@ def _get_data_type_by_hdu(hdu):
     else:
         data_type = ''
     return data_type
-
-
-def convert_spectrum1d_from_flux_to_flux_per_pixel(spectrum):
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:585]: This is now unnecessary, as Clare requested.

File: jdaviz/configs/cubeviz/plugins/tests/test_parsers.py [hunk 4]
```diff
@@ -194,15 +194,15 @@ def test_numpy_cube(cubeviz_helper):
     assert data.label == 'Array'
     assert data.shape == (4, 3, 2)  # x, y, z
     assert isinstance(data.coords, PaddedSpectrumWCS)
-    assert flux.units == 'ct'
+    assert flux.units == 'ct / pix2'
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/tests/test_parsers.py:197]: I think this fixes a bug where pure numpy cube does not get that PIX2 assigned like all the other stuff loaded into Cubeviz, right?

File: pyproject.toml [hunk 5]
```diff
@@ -23,7 +23,7 @@ dependencies = [
     "ipywidgets>=8.0.6",
     "solara>=1.40.0",
     "pyyaml>=5.4.1",
-    "specutils>=1.16",
+    "specutils@git+https://github.com/astropy/specutils.git@main",  # TODO: >=1.18.0
```
Comment (pllim) [inline pyproject.toml:26]: This requires Ricky to release specutils 0.18 first. We cannot merge this PR until that release happens.

**Comment by pllim** [comment]
@cshanahan1 , does this look about right?

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 6]
```diff
@@ -223,23 +222,22 @@ def _return_spectrum_with_correct_units(flux, wcs, metadata, data_type=None,
...
+        target_wave_unit = None
+
+    if (target_wave_unit is None) and (target_flux_unit is None):  # Nothing to convert
         new_sc = sc
+    elif target_flux_unit is None:  # Convert wavelength only
+        new_sc = sc.with_spectral_axis_unit(target_wave_unit)
+    elif target_wave_unit is None:  # Convert flux only and only PIX2 stuff
+        new_sc = sc.with_flux_unit(target_flux_unit, equivalencies=_eqv_flux_to_sb_pixel())
+    else:  # Convert both
+        new_sc = sc.with_spectral_axis_and_flux_units(
```
Comment (gibsongreen) [inline jdaviz/configs/cubeviz/plugins/parsers.py:235]: I checked the Spectrum1D class and it doesn't seem to have `with_spectral_axis_and_flux_units` method exposed. I am seeing it as in the specutils documentation in the link below. I tried loading a counts cube and when this gets called it is throwing a traceback: https://specutils.readthedocs.io/en/stable/api/specutils.spectra.spectrum_mixin.OneDSpectrumMixin.html#specutils.spectra.spectrum_mixin.OneDSpectrumMixin.with_spectral_axis_and_flux_units

**Review by cshanahan1 — APPROVED** [review]
looks good- tested it on a bunch of units loaded and it works as intended, no comments. thanks!

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 7]
```diff
@@ -522,15 +521,15 @@ def _parse_ndarray(app, file_obj, data_label=None, data_type=None,
     flux = file_obj
 
     if not hasattr(flux, 'unit'):
-        flux = flux << u.count
+        flux = flux << (u.count / PIX2)
```
Comment (gibsongreen) [inline jdaviz/configs/cubeviz/plugins/parsers.py:524]: Can we rename the `file_obj` parameter to `flux` in the method signature?

File: jdaviz/configs/cubeviz/plugins/tests/test_parsers.py (hunk 4 above)
Comment (gibsongreen) [reply jdaviz/configs/cubeviz/plugins/tests/test_parsers.py:197]: Nice catch! I just checked this on main and it wasn't being coerced

File: jdaviz/configs/cubeviz/plugins/parsers.py (hunk 2 above)
Comment (gibsongreen) [reply jdaviz/configs/cubeviz/plugins/parsers.py:558]: Good call coercing here too

**Review by gibsongreen — APPROVED** [review]
All the parser changes are concise, I tested the different input cubes for the parsers that were updated and all results are as expected, great work!

**Comment by pllim** [comment]
Thanks for the reviews!

**Pull request merged**